src/crypto/tls/handshake_client.go

```diff
@@ -146,30 +146,38 @@ func (c *Conn) makeClientHello(minVersion uint16) (*clientHelloMsg, clientKeySha
 			hello.cipherSuites = append(hello.cipherSuites, defaultCipherSuitesTLS13NoAES...)
 		}
 
-		curveID := config.curvePreferences()[0]
-		if scheme := curveIdToCirclScheme(curveID); scheme != nil {
-			pk, sk, err := generateKemKeyPair(scheme, curveID, config.rand())
-			if err != nil {
-				return nil, nil, fmt.Errorf("generateKemKeyPair %s: %w",
-					scheme.Name(), err)
-			}
-			packedPk, err := pk.MarshalBinary()
-			if err != nil {
-				return nil, nil, fmt.Errorf("pack circl public key %s: %w",
-					scheme.Name(), err)
-			}
-			hello.keyShares = []keyShare{{group: curveID, data: packedPk}}
-			secret = sk
-		} else {
-			if _, ok := curveForCurveID(curveID); !ok {
-				return nil, nil, errors.New("tls: CurvePreferences includes unsupported curve")
-			}
-			key, err := generateECDHEKey(config.rand(), curveID)
-			if err != nil {
-				return nil, nil, err
+		hello.keyShares = []keyShare{}
+		curveIDs := []CurveID{config.curvePreferences()[0]}
+
+		if config.CurveGuess != nil {
+			curveIDs = config.CurveGuess
+		}
+
+		for _, curveID := range curveIDs {
```

Comment:
Should we also check here that the CurveGuess list is a subset of CurvePreferences?

Reply:
Good question. Let me check the spec.

Also we should check for duplicates.

Reply:
Updated.